# Introduction of TICHR

<div style="padding: 15px; border: 1px solid transparent; border-color: transparent; margin-bottom: 20px; border-radius: 4px; color: #8a6d3b;; background-color: #fcf8e3; border-color: #faebcc;">
<b>We are submitting the TICHR paper for publication and welcome any feedback to improve this multiomics integration tool.</b>
</div>



## Background

<img src="_static/logo.png" style="zoom:30%;" />

Deciphering transcriptional regulation across multiple omics layers is essential for understanding cellular processes and disease mechanisms, yet remains challenging due to the limited direct connections among epigenomic, 3D genomic, and transcriptomic data. Here, we present TICHR (Transcriptional regulation analysis by integrating ChIP-seq, Hi-C and RNA-seq), a scalable multiomics integration framework that quantifies both site-to-gene and gene-level regulation through diverse weighting strategies, enabling genome-wide characterization of regulatory programs.

<img src="_static/introduction/001.png" style="zoom:80%;" />

TICHR offered sophisticated downstream functions to address diverse research questions for transcriptional regulation, including enhancer prediction, attribution of transcriptional changes, assessment of regulation–transcription concordance, identification of context-specific regulations, cross-sample analysis of large-scale and single-cell data, and characterization of time-series transcriptional dynamics. These multi-task capabilities establish TICHR as an efficient framework for studying complex transcriptional mechanisms.

-----

## Install

TICHR is developed based on Python 3.11.5. In theory, any Python version newer than 3.11.5 should be compatible. However, if you encounter any issues, we recommend creating an isolated environment with Python 3.11.5 using Conda.

### Requirement

Python packages:

-   numpy
-   pandas
-   scipy
-   matplotlib
-   seaborn
-   pyBigWig
-   scikit-learn
-   hic-straw
-   statsmodels
-   pyranges
-   tqdm
-   adjustText
-   umap-learn

Optional

- rpy2
- joblib
- multiprocessing

### Install

You can install the latest version of TICHR <u>from PyPI using pip</u>:

``` shell
pip install tichr
```

Alternatively, you can install TICHR from the <u>source code</u> on GitHub:

``` shell
git clone https://github.com/wangjk321/Tichr
cd Tichr
python setup.py install
```

## Usage and tutorial

There are two ways to use TICHR

1. **Command Line Interface (CLI)** — This is the most straightforward method for most users. After installation, you can check the available commands with:

``` shell
tichr --help
```

The current CLI provides functionality for the two core purposes of TICHR.You can see the descriptions [here](./CLI.md).

- Compute regulatory scores: site-to-gene–level RgX values; gene-level Rg values.
- Identify context-specific repressive regulation


2. **Python Module** — Advanced users can import TICHR as a Python package and use its functions programmatically. For example:

``` python
from tichr import tichr
from tichr import siteToGene
```

The python package include all TICHR functions.

### Deposited Data

Test data used in this document can be found at **Zenodo** https://zenodo.org/records/17629590 

### Workflow

**Although the individual functions of TICHR were described separately, they could be streamlined into a typical workflow**：

<img src="_static/introduction/002.png" style="zoom:80%;" />

You can find detailed usage instructions for them in the following chapters.