# 1. Compute Rg and RgX

**Computing Rg and RgX is a core functionality of Tichr.**

TICHR integrates epigenomic signals (ChIP-seq, CUT&Tag, ATAC-seq), chromatin 3D interactions (Hi-C, etc.), and gene expression data (RNA-seq) to quantify regulatory strength at both the **site-to-gene level (RgX)** and the **gene level (Rg)**, enabling genome-wide inference of transcriptional regulation.

TICHR applied multiple weighting strategies (Hi-Câ€“based, distance-based, and user-defined) to simultaneously represent the correspondence between regulatory loci and target genes (RgX) and the overall regulatory effects on each gene (Rg). The detailed methods could be found in the TICHR paper.

## A quick start
A quick example for computing Hi-C-based Rg and RgX. You can jump to the step-by-step tutotal [here](#Step-by-Step-tutorial) and the full parameter descriptions [here](#full-functions-detail).

Data could be downloaded from [here](./0.Introduction.md#usage-and-tutorial). We have prepared an example datasets at `Calculate` directory.


``` python
# import pacakge
from tichr import tichr

# Assign input file
datadir="./Data/Calculate"
candidateSite=f"{datadir}/candidatePeak.bed"
chipfile=f"{datadir}/H3K27ac.bigwig"
gtfile=f"{datadir}/hg38_gt.tsv"
genebedfile=f"{datadir}/tichr_allcoding_hg38.tsv"

# Generating candidate sites
tichrobj = tichr.Tichr(candidateSite,[chipfile,],gtfile,genebedfile,peakToGeneMaxDistance=100000)
tichrobj.makeSiteBed()
tichrobj.makeSiteBdg("coverageBed",file_type="bigWig")

# Generating Hi-C weights
tichrobj.hicfilepath = "../ENCFF621AIY.hic"
tichrobj.proceessHiC(50000,'rawhic_sparse',"VC_SQRT",threads=12,contactNorm='default')

# Compute RgX and Rg
tichrobj.computeAllGene("hic")

# Output files
tichrobj.save(outname="test",header=False)
tichrobj.clean()
```

These commands will generate two output files containing the gene regulatory potential (Rg) and site-to-gene regulatory strength (RgX) values, respectively.

```
***Checking file availablity
***Temporary directory created at: tichr_tmp_dcFdUuaogE
***Setting candidate sites
......Use a user-defined candidate bed file
***Extracting read counts from epigenome files
......Input epigenome signal is Bigwig.
***Processing hic ...
......Using 12 threads to process Hi-C. More threads require larger memory.
......Normalizing the contact weight.
......Finished the Hi-C weight process.
***Computing RgX and Rg
***Using hic mode
***Processing genes: 100%|â–ˆ| 20049/20049
***Finished
***Save RgX and Rg to tsv tables.
......Saved to test_RgxDf.tsv.gz and test_RgDf.tsv.gz
clean all attribute to release memory.
```

> TICHR is designed in an object-oriented manner, allowing you to modify any parameter (attribute) at any stage of the workflow.

## Step-by-Step tutorial 

In this section, we present a complete example workflow illustrating the essential steps for computing RgX values for all site-to-gene regulatory links and Rg values for each gene.


## Step1. Prepare input and Create Tichr object

Tichr is flexible for various input formats and research purposes. The following section lists the typical input files required.


``` python
from tichr import tichr

#data path
datadir="./Data/Calculate"
#reference
gtfile=f"{datadir}/hg38_gt.tsv"
refGeneFile=f"{datadir}/tichr_allcoding_hg38.tsv"
#sites and genes
candidatesite=f"{datadir}/candidatePeak.bed"
candidateGeneFile=f"{datadir}/candidateGene.tsv"
#epigenome signal
DNase_rep1=f"{datadir}/DNase_rep1.bigwig"
DNase_rep2=f"{datadir}/DNase_rep2.bigwig"
readFileList=[DNase_rep1,DNase_rep2,]

H3K27ac=f"{datadir}/H3K27ac.bigwig"
readFileList2=[H3K27ac,]
#Hi-C
hicfilepath="../ENCFF621AIY.hic"
#Create TICHR object
tichrobj = tichr.Tichr(candidatesite,readFileList,gtfile,candidateGeneFile,
                      refGeneFile=refGeneFile,ifTSSrange=500,S2Gmax=100000,
                      hicfilepath=hicfilepath,readFileList2=readFileList2)
```

#### Required parameters

- `candidatesite`: Candidate regulatory sites, 
  - this could be a BED3 file (e.g, `"testdata/candidatePeak.bed"`), 
  - or a predefined type specified by a string: `"denovo_peak"`,`"surronding_bin"`,`"onlypromoter"`. 
  - Also see [below](#step2-generate-candidate-sites) 


- `readFileList`: **A list** of input files for epigenomic data such as ChIP-seq, ATAC-seq, or CUT&Tag. 
  - Supported formats include BAM, BigWig, and BedGraph. 
  - Multiple files should be provided as a list, e.g., `["testdata/DNase_rep1.bam","testdata/DNase_rep2.bam"]`
  <div style="padding: 10px; border: 1px solid transparent; border-color: transparent; margin-bottom: 1px; border-radius: 10px; color: #000000ff;; background-color: #b4d9bfff; border-color: #faebcc;">
  <em>It must be provided as a list. Even if you have only one file, please specify it as ["testdata/DNase_rep1.bam"].</em>
  </div>


- `gtfile`: A tab-separated genome table file, with 
    - column 1 specifying chromosome names (e.g., chr14)
    - column 2 indicating chromosome lengths (e.g., 107043718).

- `candidateGeneFile`: A tab-separated file listing candidate genes. It must contain at least six columns in the following order:   
    - chromosome
    - start
    - end
    - gene symbol
    - gene ID
    - strand (+/-)].

#### Optional parameters

- `refgene_file`:  Reference gene file in the same format as `candidateGeneFile`. 
  - This file is used to define gene promoter regions. 
  - If you do not provide one, TICHR will use the `candidateGeneFile` to define promoters. 
  - If you need promoter definitions for other genes, please specify a separate file.

- `ifTSSrange`: Defines the promoter region as transcription start site (TSS) Â± this range. 
  - Type: int. Default: 500.

- `S2Gmax`: Maximum distance (in base pairs) allowed between a peak and a gene for linking. 
  - Type: int. Default: 100000. 
  - A large value will generate a longer calculation time and the size of output files. 
  - For most cases, 500,000 bp (500 kb) is recommended. 
  - A window of 1,000,000 bp (1 Mb) will include most potential regulatory interactions
  - 5 Mb can be used to capture ultraâ€“long-range interactions.

- `hicfilepath`: Path to hic files in Juicer `.hic` format.
  - The juicer .hic format will give the best performance.
  - Alternatively, if you do not have a Juicer-format contact map, 
    - you may provide a dense matrix (compressed) of intra-chromosomal contacts. 
    - An example file can be found at [HiC1Dmetrics](https://github.com/wangjk321/HiC1Dmetrics/blob/master/test_data/GSE104334_Ctrl.chr21.matrix.gz).  
    - Matrices in this format can be generated by `h1d` (see the documentation [here](https://h1d.readthedocs.io/en/latest/basic.html#make-contact-matrix)).
    - Also check [here](#full-parameters)


- `readFileList2`:  A second set of epigenome data files, in the same format as readFileList. 
  - For example, DNase signals can be provided in readFileList and H3K27ac signals in readFileList2. 
  - These two signals are combined using the geometric mean


------

## Step2. Generate candidate regulatory sites

``` python
encodeblack="testdata/hg38-blacklist.v2.bed"
tichrobj.makeSiteBed(macs2species='hs',binResolution=100,
                blackregion=encodeblack,tmpdir=None,fixPeakWidth=False)
```

<div style="padding: 15px; border: 1px solid transparent; border-color: transparent; margin-bottom: 20px; border-radius: 10px; color: #000000ff;; background-color: #b5cabaff; border-color: #faebcc;">
You can check the generated sites file via the attribute `tichrobj.candidatesite_file`.
</div>



#### Four different modes

**Depending on the `candidatesite` parameter (as described above), a list of putative regulatory regions can be generated automatically or supplied by the user.**

> 1. **BED3 file**: User-provided BED regions. The input BED intervals are resized to 500 bp centered windows and filtered to remove blacklisted regions.  
> 2. **`"denovo_peak"`**: De novo peaks. Peaks are called from the input BAM files, with blacklisted regions excluded.  
> 3. **`"surronding_bin"`**: Surrounding bins. Fixed-size genomic bins (e.g., 100 bp) tiled around gene TSSs based on annotation.  
> 4. **`"onlypromoter"`**: Promoter-only. Â±100 bp regions centered on the TSS.

If you use user-defined sites, please run:
``` python
# For most cases,run
tichrobj.makeSiteBed()
# if you want to specify parameter, use
tichrobj.makeSiteBed(blackregion=encodeblack,tmpdir=None,fixPeakWidth=True)
```

Otherwise, please run:
``` python
tichrobj.makeSiteBed(macs2species='hs',binResolution=100,
                blackregion=encodeblack,tmpdir=None,
                fixPeakWidth=True,only_promoter_area=100)
```

#### Optional parameters

- `macs2species`: Used only when `candidateSite` is set to `"denovo_peak"`. 
  - Specifies the effective genome size for MACS2 peak calling. 
  - It can be a numeric value (e.g., 1000000000) or a shortcut string ('hs' for human, 'mm' for mouse, 'ce' for C. elegans, 'dm' for Drosophila). 
  - Type: str, default: "hs".

- `binResolution`: Used only when candidateSite is set to `"surronding_bin"`. 
  - Defines the bin size (in base pairs) for creating windowed candidate sites. 
  - Default: 100.

- `fixPeakWidth`: Applicable only for given `user-defined-bed3` candidate sites.
  - If set to True, each peakâ€™s width is fixed to 500 bp by centering and extending Â±250 bp.
  - default: True

- `only_promoter_area`: Only be used when `candidatesite` is set to be â€œonlypromoterâ€. 
  - Default: 100

- `blackregion`: Regions to exclude, such as ENCODE blacklist sites, provided in BED3 format.

- `tmpdir`: Temporary directory name for intermediate files. 
  - Default is a randomly generated name like tichr_tmp_rsDuchihKJ.

The output information will be

``` text
***Temporary directory created at: tichr_tmp_BhmttQjTGC
***Setting candidate sites
......Use a user-defined candidate bed file
```

---

## Step3. Quantify epigenome signals

#### General usage
In this step, TICHR calculates the activity of each candidate site based on the input epigenomic signals, such as ChIP-seq, ATAC-seq, or CUT&Tag data. This process includes four components:

> 1. Coverage calculated via read counting or signal averaging.  
> 2. Optional SPMR normalization.  
> 3. Replicates merged by mean or sum.  
> 4. *(Archived)* Quantile normalization to reference profiles is supported.


``` python
# Generally, you can directly use
tichrobj.makeSiteBdg(file_type="bigWig")

# The full parameters are
tichrobj.makeSiteBdg(coverageMethod="coverageBed",spmr = True,
                    multiBAMmerge='mean',file_type="bigWig")

```

- **`coverageMethod`**: Method used to compute coverage.  
  - For most users, `"coverageBed"` is recommended.  
  - For BAM input, you may also use `"pileup"` to generate epigenomic coverage using the MACS2 pileup function.

- **`spmr`**: Whether to normalize signal by total mapped reads (Signal Per Million Reads).  
  - Set this to `True` if you plan to compare Rg or RgX across samples. 
  - However, note that SPMR normalization may not always be appropriate, because global signal differences may reflect true biological variation, as discussed in the paper.

- **`multiBAMmerge`**: Strategy for merging multiple replicates.  
  - Options: `"mean"` (default) or `"sum"`.

- **`file_type`**: Format of the epigenomic input files.  Supported types:
  - `"Bam"`
  - `"bigWig"`
  - `"bedGraph"`.

The output information:

``` text
***Extracting read counts from epigenome files
......Input epigenome signal is Bigwig.
***Extracting read counts from epigenome files
......Input epigenome signal is Bigwig.
```


<div style="padding: 15px; border: 1px solid transparent; border-color: transparent; margin-bottom: 20px; border-radius: 10px; color: #000000ff;; background-color: #b5cabaff; border-color: #faebcc;">
<em>The main obtained attributte is `tichrobj.candidatesite_coverage`, which contains the following information:</em>
</div>

``` python
tichrobj.candidatesite_coverage.head()

```


| Chromosome | Start Position | End Position | Value                                                        | ifPromoter |
| ---------- | -------------- | ------------ | ------------------------------------------------------------ | ---------- |
| chr1       | 25933091       | 25934590     | 341.499634                                                   | 0          |
| chr1       | 26378064       | 26378863     | 552.494344                                                   | 0          |
| chr1       | 26508509       | 26508806     | 21.633308                                                    | 1          |
| chr1       | 27867080       | 27867532     |  123.405222| 0          |


#### Quantile normalization

**Important notice: Quantile normalization is generally NOT recommended**.
- This is the same quantile-normalization procedure for epigenomic signals described in the ABC Score framework.
- However, it typically provides little improvement and may even obscure biologically meaningful global signal differences.
- The available reference file is limited to a K562-specific empirical distribution for only **H3K27ac** and **DNase**.
- Because such predefined references are unavailable for most cell types, this approach is not suitable for general use.


> Quantile normalization adjusts region-wise signal intensities by ranking them, mapping their empirical distribution to a reference profile using linear interpolation, and updating values accordingly.

``` python
# if you really want to use:

tichrobj.makeSiteBdg(file_type="bigWig",quantileref="EnhancersQNormRef.K562.txt",
                     quantile_method="rankpercent",separatepromoter=False,
                     signaltype="DNase",signal2type="H3K27ac")
```

**Parameters**

- `quantileref = None`  
  Path to a quantile-normalization reference file.  
  An example file can be found [here](https://github.com/wangjk321/TICHR/blob/main/tichr/supportcode/EnhancersQNormRef.K562.txt.gz), which is a K562-specific empirical reference derived from the ABC paper.

- `quantile_method = None`  
  Strategy for using the reference. Options:  
  - `'rankpercent'` â€” use the *quantile* column in the reference  
  - `'rank'` â€” use the *rank* column in the reference

- `signaltype = None`  
  Signal type to normalize. Supported types: `"H3K27ac"` or `"DNase"`.

- `separatepromoter = False`  
  Whether to normalize promoter and non-promoter regions separately.  
  If `True`, promoters and non-promoters use different reference distributions.

- `signal2type = None`  
  Required only if `readFileList2` is provided above.  
  Specifies the signal type for the second dataset. Must be `"H3K27ac"` or `"DNase"`.


---

## Step4. Generate Site-to-Gene Weight

#### Description

**This step is required *only* when using Hi-Câ€“based weighting.  


TICHR provides two strategies to compute the weight between Site-to-Gene pairs:

1. **Distance-based weighting functions**
2. **Hi-Câ€“informed weighting**

For detailed methodology and comparisons between these approaches, please refer to our paper.


Distance-based functions include 8 types: 

> - `Sigmoid`
> - `Exponential`
> - `Powerlaw`
> - `NormPL`
> - `Linear`
> - `Constant`
> - `Closest`
> - `OnlyPromoter`

Hi-C-informed weighting include 5 types:

> - `default`
> - `oe`
> - `abc`
> - `0to1`
> - `total`

<div style="padding: 15px; border: 1px solid transparent; border-color: transparent; margin-bottom: 20px; border-radius: 10px; color: #000000ff;; background-color: #eeeee8ff; border-color: #faebcc;">
If you choose a distance-based weighting function (e.g., `Exponential`), you may skip this step.**
</div>

For each chromosome in the analysis set, Hi-C contact matrices were extracted and processed using the `gethicfile` function. 

#### Full parameters

``` python
tichrobj.proceessHiC(hicRes,hicDataType,hicNormType,juicertool=None,
                    threads=8,contactNorm='default')
```

- **`hicRes`**: Resolution of the Hi-C contact matrix.  
  - `10,000` is recommended for most use cases.  
  - Higher resolutions (e.g., `5,000`, `1,000`) require more memory and CPU time.  
  - Lower resolutions (> `50,000`) run faster but provide coarser signals.

- **`hicDataType`**: Workflow depends on the Hi-C input format (dense / raw dense / raw sparse).  
  **In most cases, `'rawhic_sparse'` is recommended. Provide a `.hic` file in Step 1.**
  - **`rawhic_sparse`**: For sparse extraction from `.hic` using the hic-straw API. Recommended!
  - **`rawhic_dense`**: 
    - Suitable for â€œnon-standardâ€ `.hic` files (e.g., generated by JuicerTools > 2.0). 
    - For raw `.hic` input. Contact matrices are exported chromosome-wise using Juicer at the specified resolution
    - Then normalized with a specific method.
  - **`matrix_dense`**: For precomputed dense matrices (one per chromosome). 
    - User-defined matrices are directly loaded and normalized.
    - An example file can be found at [HiC1Dmetrics](https://github.com/wangjk321/HiC1Dmetrics/blob/master/test_data/GSE104334_Ctrl.chr21.matrix.gz).  
    - Matrices in this format can be generated by `h1d` (see the documentation [here](https://h1d.readthedocs.io/en/latest/basic.html#make-contact-matrix)).

- **`hicNormType`**: Initial normalization method for Hi-C matrices, such as `VC_SQRT`, `KR`, `SCALE`, etc.

- **`contactNorm`**: Additional user-defined normalization strategy for Hi-Câ€“informed weighting:  
  1. `default`: Default TICHR normalization  
  2. `abc`: ABC-style distance decay correction  
  3. `oe`: Observed/Expected (OE) normalization  
  4. `0to1`: Rescale to \[0, 1\] using the 95th percentile as the max  
  5. `total`: Total-sum normalization (divide by matrix sum and multiply by `1e7`)

- **`threads`**: Number of threads to process Hi-C files.  
  Effective only for `rawhic_sparse` and `rawhic_dense`, where parallelization is done across chromosomes.

- **`juicertool`**: Used only in `rawhic_dense` mode.  
  If your `.hic` file cannot be processed by the built-in JuicerTools version, you may supply a custom JuicerTools JAR (see: https://github.com/aidenlab/JuicerTools).


#### Usage example

1. recommended `rawhic_sparse` mode.

``` python
tichrobj.proceessHiC(50000,'rawhic_sparse',"VC_SQRT",
                   threads=12,contactNorm='default')
```

The output information:
``` text
***Processing hic ...
......Using 12 threads to process Hi-C. More threads require larger memory.
......Normalizing the contact weight.
......Finished the Hi-C weight process.
```

<div style="padding: 15px; border: 1px solid transparent; border-color: transparent; margin-bottom: 20px; border-radius: 10px; color: #000000ff;; background-color: #b5cabaff; border-color: #faebcc;">
The main obtained information is `tichrobj.nomhicdf`, which provide the contact information for each chromosome.
</div>

``` python
tichrobj.nomhicdf["chr19"]
```

| Bin Pair       | posX   | posY   | Counts      |
| -------------- | ------ | ------ | ----------- |
| 50000to50000   | 50000  | 50000  | 2477.239014 |
| 200000to200000 | 200000 | 200000 | 1096.694336 |
| 50000to250000  | 50000  | 250000 | 52.562267   |
| ...  | ...  | ... | ...  |

2. `rawhic_dense` mode.

``` python
# You can change the hic file at any time.
tichrobj.hicfilepath = "../ENCFF621AIY.hic"
# Run dense mode. This will also generating a juicerdump directory 
# that contains all the contact matrices.
tichrobj.proceessHiC(50000,'rawhic_dense',"VC_SQRT",
                   threads=12,contactNorm='default',
                     juicertool="juicer_tools.2.13.05.jar")
```

The output information:
``` text
***Processing hic ...
......Using 12 threads to process Hi-C. More threads require larger memory.
......Dumping contact matrix from .hic file ......

......The dump command information is in juicer_dump_info.txt.
......If you have any error, please check that file at first......

......Dump finished, output is in ./juicerdump
......Normalizing the contact weight.
......Finished the Hi-C weight process.
```

3. `matrix_dense` mode.

``` python
# For the matrix_dense mode, you should give a directory as Hi-C file path
tichrobj.hicfilepath = "./juicerdump/50000/"
# If you matrix file is "./juicerdump/50000/observed.VC_SQRT.chr1.matrix.gz"
# The hicprefix="observed.VC_SQRT.",hicsuffix=".matrix.gz"
tichrobj.proceessHiC(50000,'matrix_dense',"VC_SQRT",contactNorm='default',
                    hicprefix="observed.VC_SQRT.",hicsuffix=".matrix.gz")
```


## Step5. Calculate Rg and Rgx

#### Required parameter

Only one required paramter is `weightType`. it could be
- `hic`
- `fixedFunction`
- `userWeight`

1. Hi-C-based weighting

``` python
tichrobj.computeAllGene("hic")
```

2. Distance-based function

``` python
tichrobj.computeAllGene("fixedFunction", fixedFunctionType='exponential')
```

If you use this, you must give a specific type (`fixedFunctionType`) from
> - `Sigmoid`
> - `Exponential`
> - `Powerlaw`
> - `NormPL`
> - `Linear`
> - `Constant`
> - `Closest`
> - `OnlyPromoter`

3. User-defined weight

``` python
tichrobj.computeAllGene("userWeight", userWeightFile='user_weight.tsv')

```

#### Output results

The successful running information
``` text
***Computing RgX and Rg
***Using fixedFunction mode
***Processing genes: 100%|â–ˆ| 2412/2412 [00:12<00:00, 196.6
***Finished
```

Then it will generate a site-to-gene level RgX and gene-level Rg. Only sites within a maximum distance from the gene are considered.

<div style="padding: 15px; border: 1px solid transparent; border-color: transparent; margin-bottom: 20px; border-radius: 10px; color: #000000ff;; background-color: #b5cabaff; border-color: #faebcc;">
The main obtained information is `tichrobj.RgxDf`:
</div>

``` python
tichrobj.RgxDf.head()
```

| Peak Chr | Peak Start | Peak End | Activity   | Gene Symbol       | Gene Chr | Gene Start | Gene End | Strand | Gene ID           | Weight     | Rgx Raw Value | Rgx Percent |
| -------- | ---------- | -------- | ---------- | ----------------- | -------- | ---------- | -------- | ------ | ----------------- | ---------- | ------------- | ----------- |
| chr1     | 25933091   | 25934590 | 341.499634 | ENST00000516882.1 | chr1     | 26006216   | 26006216 | +      | ENSG00000252691.1 | 984.895508 | 336341.5      | 1.0         |
| chr1     | 26378064   | 26378863 | 552.494344 | ENST00000319041.6 | chr1     | 26280122   | 26280122 | +      | ENSG00000142669.9 | 1761.7518  | 973357.9      | 0.077389    |
| chr1     | 26378064   | 26378863 | 552.494344 | ENST00000492808.1 | chr1     | 26317958   | 26317958 | +      | ENSG00000169442.4 | 2596.0479  | 1434302.0     | 0.893021    |

<div style="padding: 15px; border: 1px solid transparent; border-color: transparent; margin-bottom: 20px; border-radius: 10px; color: #000000ff;; background-color: #b5cabaff; border-color: #faebcc;">
and `tichrobj.RgDF` (The input gene file with the last column represent the Rg values):
</div>

``` python
tichrobj.RgDf.head()
```

| Chromosome | Start    | End      | Symbol | Gene ID            | Strand | Value         |
| ---------- | -------- | -------- | ------ | ------------------ | ------ | ------------- |
| chr1       | 25272548 | 25272548 | geneA  | ENSG00000187010.14 | +      | 0.000000      |
| chr1       | 26006216 | 26006216 | geneB  | ENSG00000252691.1  | +      | 336341.455414 |
| chr1       | 26183204 | 26183204 | geneC  | ENSG00000142675.13 | +      | 0.000000      |


You can save the main output by 

``` python
tichrobj.save(outname="test", header=False)
```

- `outname`: output prefix
- `header`: if output with column names.


Running information:
``` text
***Save RgX and Rg to tsv tables.
......Saved to test_RgxDf.tsv.gz and test_RgDf.tsv.gz
```

Finally, please clean the tichr object

``` python
tichrobj.clean()
```

#### Full parameters

``` python
tichrobj.computeAllGene(weightType,
                fixedFunctionType='exponential',halfDistance=10000,
                 setpromoter1=False,threads=1, 
                 userWeight=Noneï¼Œ noise_ratio=0,noise_quantile=0):
```

- `weightType`: Determines how the site-to-gene weight is calculated. Options:
  - `"hic"` : based on Hi-C contact frequency.
  - `"fixed_function"` :based on distance-based function.
  - `"userWeight"` : customized weight file by user.

- `fixedFunctionType`: Specifies the function used if weightType="fixed_function". Options include: 
  - "Sigmoid", "Exponential", "Powerlaw", "NormPL", "Linear", "Constant", "Closest", or "OnlyPromoter".
  - default: "Exponential"

- `halfDistance`:  Distance (in bp) at which the weight decays to 0.5 
  - work for supported functions [sigmoid,exponential,powerlaw,linear-half].
  - default: 10000

- `setpromoter1`: If it's True, sets the RgX weight of promoter regions to 1.

- `threads`: Number of threads for calculation. Not recommended for Hi-C mode because it consume much memory (some times even slower with multiple threads). 

- `goldWeightDf:` File path (.tsv) for the user-defined weights.

| chr1 | 8080000   | 8090000   | chr1 | 8370000   | 8380000   | 64 |
|------|-----------|-----------|-------|-----------|-----------|-------|
| chr1 | 17280000  | 17290000  | chr1 | 17330000  | 17340000  | 45 |
| chr1 | 20000000  | 30000000  | chr1 | 20000000  | 30000000  | 55 |
| chr1 | 38270000  | 38280000  | chr1 | 38320000  | 38330000  | 26 |
| chr1 | 42030000  | 42040000  | chr1 | 42500000  | 42510000  | 3  |
| chr1 | 42270000  | 42280000  | chr1 | 42500000  | 425| 23|

<div style="padding: 15px; border: 2px solid transparent; border-color: transparent; margin-bottom: 20px; border-radius: 0px; color: #FAFAFA;; background-color: #000000B6; border-color: #000000FF;">
Low-signal noise can be reduced by filtering S2G links using thresholds such as an RgX_ratio > 0.01 or a raw RgX value above the 0.01 quantile prior to Rg calculation.
</div>

- `noise_ratio`:
  - If a gene's RgX value proportion relative to its Rg value (i.e., RgX_ratio) is less than this threshold,
  - the RgX value will be set to 0.
  - Default: 0. Recommended: 0.01.

- `noise_quantile`:
  - If a gene's RgX value falls below this percentile among all its RgX values,
  - the RgX value will be set to 0.
  - Default: 0. Recommended: 0.01.



<details>

<summary> Archived parameter (usually not used)  (<u>show details</u>)</summary> 

- given_gamma
- given_scale
- ref_gamma
- ref_scale
- hicmindistance
- ifUseHiCRef

</details>

----

## Step6. RgX and Rg output.


#### Manipulate Tichr object


<u> Tichr follows an object-oriented programming design, allowing you to easily reassign parameters at intermediate steps and rerun the analysis as needed.</u>

For example, if you want to use a new Hi-C file, you can do:

``` python
tichrobj.hicfilepath = new_hic_file
tichrobj.proceessHiC(50000,'rawhic_sparse',"VC_SQRT")
tichrobj.computeAllGene("hic")
```

Another important note: when using ultra-high-resolution Hi-C data or a large number of peak-to-gene associations, the Tichr object may consume a significant amount of memory. You can check the memory usage as follows:


``` python
def format_bytes(size):
    for unit in ['B', 'KB', 'MB', 'GB', 'TB']:
        if size < 1024:
            return f"{size:.2f} {unit}"
        size /= 1024

from pympler import asizeof
size_bytes = asizeof.asizeof(tichrobj)
print("Raw size (bytes):", size_bytes)
print("Readable size:", format_bytes(size_bytes))
```

Example output:

``` text
Raw size (bytes): 26964769808
Readable size: 25.11 GB
```

To clear major memory-consuming attributes (such as Hi-C matrices and intermediate results), you can call:

``` python
# clean all
tichrobj.clean()

# clean only tmp directiory
tichrobj.clean(onlytmp=True)
```


Then youâ€™ll see the memory has been released:

``` text
Raw size (bytes): 1387440
Readable size: 1.32 MB
```

----

#### What is RgDf and RgxDf?

RgDf and RgxDf are two primary outputs of Tichr, both of which are objects of the Pandas DataFrame class. They record gene-level data and gene-site-level data, respectively.

RgDf contains 10 columns with the following content:

- Column 1: gene chr
- Column 2: gene start
- Column 3: gene end
- Column 4: gene ID
- Column 5: gene symbol
- Column 6: gene strand
- Column 7: gene Rg score  -  the total regulatory potential of the gene

RgxDf consists of 13 columns, which are:

- Column 1: site chr
- Column 2: site start
- Column 3: site end
- Column 4: site epigenome signal â€” indicates the intensity of the epigenome signal at this site
- Column 5: gene ID
- Column 6: gene chr
- Column 7: gene start
- Column 8: gene end
- Column 9: gene strand
- Column 10: gene symbol
- Column 11: site-to-gene weight â€” the weight of the site-to-gene link, used for multiplication with the site epigenome signal
- Column 12: Rgx score â€” the product of site-to-gene weight and site epigenome signal
- Column 13: Rgx ratio â€” the proportion of the Rgx value of this site-to-gene link relative to the Rg value of the corresponding gene

----

#### Filter regulations

You can also re-generate a new RgX and Rg files by changing the maximum site-to-gene distance. For example, if you obtained a TICHR output with S2Gmax=500000, you can easily re-generate new TICHR output by:

Assume the input files:
``` python
rgxfile_raw = "test_RgxDf.tsv"
rgfile_raw = "test_RgDf.tsv"
```

Import required libraries and initialize the FilterTichr object. If the files do not have headers, use header=False; if they have headers, use header=True.

``` python
from tichr.siteToGene import *
f = postTichr(rgxfile_raw, rgfile_raw, header=False)
```

Filter S2G links by distance. For example, keep only links where the peak-to-TSS distance is less than 100000. The filtered RgxDf is stored in f.df and the updated RgDf is stored in f.rgdf.

``` python
f.cut_distance(maxdis=100000)
print(f.df.head())
print(f.rgdf.head())
```

Next, filter the links based on signal and ratio. `minRaw` specifies the percentile threshold for RgX_raw to remove very low signals, and `minRatio` specifies the minimum value of `RgX_ratio` to remove very small ratios. The filtered results are updated in f.df and f.rgdf.

``` python
f.cut_score(minRaw=0.01, minRatio=0.01)
df_filtered = f.df
rgdf_filtered = f.rgdf

```

Finally, save the filtered results to compressed TSV files. The header argument controls whether to include column names; if not specified, it defaults to the input file format.

``` python
f.save(outname="filtered", header=False)

```
After running the above steps, you will have two output files: filtered_RgxDf.tsv.gz containing the filtered enhancer-gene links and filtered_RgDf.tsv.gz containing the corresponding Rg values. This workflow handles initialization, distance filtering, signal and ratio filtering, and saving in a complete pipeline.


----

## Step7. Adjust S2G regulation

TICHR can optionally perform adjustments according to the following rules. For more decription, see our original paper:

1. `adjRaw`: **Raw regulatory values instead of ratios**. The raw RgX value is used instead of the within-gene ratio, which is the default behavior in Tichr. Note the Rg values can only be calculated by RgX_raw but not RgX_ratio.
2. `adjrank`: **incorporating gene expression concordance**. For each geneâ€“site pair ð‘–, the RgX value is scaled by a penalty term that reflects the absolute rank difference between RgX and gene expression (TPM) across all pairs. 
3. `adjStruc`: **leveraging higher-order chromatin structures**.The final score is further weighted by the product of multiple user-defined structural factors. For each criterion (e.g., residing in the same TAD, crossing a boundary, or being connected by a loop), the score is multiplied by a user-specified factor if the condition is met; otherwise, the factor defaults to 1.

<div style="padding: 15px; border: 1px solid transparent; border-color: transparent; margin-bottom: 20px; border-radius: 10px; color: #000000ff;; background-color: #b5cabaff; border-color: #faebcc;">
<p><strong>Important:</strong></p>
<p>
The adjustment is recommended for S2G-level studies, but NOT recommended for gene-level Rg analyses.
</p>
</div>

Here is an example to demonstrate how to adjust RgX and Rg:

#### Prepare the files before adjustment.

The adjustments are performed based on the RgX and Rg files generated above.

Here, we specify the raw RgX and Rg `.tsv` data files, along with the output directory where the adjusted results will be saved.

``` python

from tichr import tichr
from tichr.siteToGene import *
rgxfile_raw="test_RgxDf.tsv.gz"
rgfile_raw="test_RgDf.tsv.gz"

outdir = "adjusted/"
import os
if not os.path.exists(outdir):
    os.makedirs(outdir)

```

#### Prepare gene TPM file

The TPM file provides gene expression values used for downstream adjustments.

Please specify the file path and required columns:
- The TPM file must contain at least two columns: one for the gene ID and one for the TPM value .
- The gene IDs must match those used in [Step 1](#step1-prepare-input-and-create-tichr-object).

``` python 

tpmFile="./Data/Calculate/K562.genes.TPM.txt" # file location
tpmGeneID=2  #Column number for gene ID (1-based)
tpmCols=[3,4,]  # TPM values for multiple replicates across columns (1-based)
tpmHead=True  # if the tpmfile contains column names, set it to true.

```

#### Prepare sturcture file

Structural features (e.g., TADs, loops) are used to assign weights during RgX adjustment.  
You need to specify the feature type, the corresponding input file, and the weight value:

``` python

strucTypeList = ["boundary","tad","loop","stripe","compartmentSame"]

strucFileList = [
    "./Data/Calculate/ENCFF621AIY/tad.boundary.point",
    "./Data/Calculate/ENCFF621AIY/tad.region",
    "./Data/Calculate/ENCFF621AIY/loop.bedpe",
    "./Data/Calculate/ENCFF621AIY/stripe.bedpe",
    "./Data/Calculate/ENCFF621AIY/pc1.compart",
]

strucWeightList = [0.5,1.2,5,2,2]

```

**strucTypeList**: One or more of the following options:  
`["boundary", "tad", "loop", "stripe", "compartmentSame"]`

**strucFileList**: Structural feature files corresponding to each entry in `strucTypeList`.  
Each file must follow a specific format:

  - `tad.boundary.point`: Two columns â€” chromosome name and the midpoint of the TAD boundary.  
  - `tad.region`: Three columns â€” chromosome name, TAD start, and TAD end.  
  - `loop.bedpe`: Six columns â€” anchor1 chromosome, anchor1 start, anchor1 end, anchor2 chromosome, anchor2 start, and anchor2 end.  
  - `stripe.bedpe`: BEDPE-like format, using the standard output from Stripenn.  
  - `pc1.compart`: Four columns â€” chromosome name, start, end, and compartment label (*compartmentA* or *compartmentB*).

**strucWeightList**: Weights assigned to each structural feature type.  
The default setting is: `[0.5, 1.2, 5, 2, 2]`.


#### Adjust for RgX and Rg

``` python

adjS2G(rgxfile_raw,rgfile_raw,tpmFile,
         strucTypeList,strucFileList,strucWeightList,
         tpmCols=tpmCols,tpmHead=tpmHead,tpmGeneID=tpmGeneID,
         rankType="sumrank",outdir="adjS2Gout",RgXhead=False)
```
Important parameters:

- `rankType`: For `adjrank`, choose either `"sumrank"` or `"diffrank"`.
- `outdir`: Output directory for the results.
- `RgXhead`: Set to `True` if the RgX files contain header lines.

This adjustment step applies a multi-factor correction to both **RgX** and **Rg** values.  
The adjusted results are saved as **adjall.rgx.tsv** and **adjall.rg.tsv** in the specified output directory.

Successful execution will display messages similar to:

``` text
***Performing adjRaw
***Performing adjStruc
......strucFile, strucType and strucWeight are all list
***Performing adjrank
......Finish the adjustment for Rg and RgX
***Saving adjusted output
```





--------------------

<details>


<summary>(Archived) adjust after matching true label </summary>

This is a old method, two points are important:

- adjust structure before calculating RgX
- adjust gene TPM after calculateing Rgx and matching to the real value

Prepare input data

``` python
import sys
sys.path.append('/home/wang/github/Tichr-CLI/tichr')
from tichr import *
from adjustRgx_old import *

candidategene="/home/wang/Tichr/2024June/predictEP_Fulco/candidate_gene.bed"
referencegene = '/home/wang/Tichr/2024June/predictEP_Fulco/referencegene.bed'
candidatesite="/home/wang/Tichr/2024May/K562_knownEP/Fulco/allsite/hg38_FulcoCandidates.bed"
golddf = "/home/wang/Tichr/datahub/EPinter/EP_CRISPRi_screens/hg38/Fulco_interactions_wScore.txt"

bamfile1="/home/wang/Tichr/datahub/EPinter/K562_NT_H3K27ac+DNase/DNase/hg38_wgEncodeUwDnaseK562AlnRep1.sorted.bam"
bamfile2 = "/home/wang/Tichr/datahub/EPinter/K562_NT_H3K27ac+DNase/DNase/hg38_wgEncodeUwDnaseK562AlnRep2.sorted.bam"
bamfile3 = "/home/wang/Tichr/datahub/EPinter/K562_NT_H3K27ac+DNase/ENCFF600THN.bam"

gtfile="/home/wang/database/hg38/genometable_all.txt"
rawhic="/home/common/rawdata/Hi-C_juicer_file/K562/ENCFF621AIY/ENCFF621AIY.hic"
gt="~/database/hg38/genome_table"

```


Calculate Rgx without any adjustment

``` python
tichr_obj_noquantile = Tichr(candidatesite,[bamfile1,bamfile2],gtfile,
                  candidategene,refgene_file=referencegene,
                    ifTSSrange=500,peakToGeneMaxDistance=500000,
                  hicfilepath=rawhic,readFileList2=[bamfile3,])

tichr_obj_noquantile.makeSiteBed(macs2species='hs',binResolution=100,tmpdir='tmp_K562fulco')
tichr_obj_noquantile.makeSiteBdg("coverageBed",spmr = False)

tichr_obj_noquantile.hicfilepath="/home/common/rawdata/Hi-C_juicer_file/K562/ENCFF621AIY/ENCFF621AIY.hic"
tichr_obj_noquantile.proceessHiC(25000,"rawhic_sparse","VC_SQRT",contactNorm='no_further',threads=12)
tichr_obj_noquantile.computeAllGene("hic",ifUseHiCRef=False)

predictdfname= "noStructure"
tichr_obj_noquantile.RgxDf.to_csv("resultdf_Fulco_2025/"+predictdfname+"_Rgx_beforematch.tsv",
                                  index=None,sep="\t")
tichr_obj_noquantile.RgDF.to_csv("resultdf_Fulco_2025/"+predictdfname+"_Rg_beforematch.tsv",index=None,sep="\t")

```

Calculate Rgx with structure adjustment

``` python
structureTypeList = ["boundary","tad","loop","stripe","compartmentSame"]
structureFileList = [
    "/home/wang/Tichr/datahub/EPdata/K562/ENCFF621AIY/tad.boundary.point",
    "/home/wang/Tichr/datahub/EPdata/K562/ENCFF621AIY/tad.region",
    "/home/wang/Tichr/datahub/EPdata/K562/ENCFF621AIY/loop.bedpe",
    "/home/wang/Tichr/datahub/EPdata/K562/ENCFF621AIY/stripe.bedpe",
    "/home/wang/Tichr/datahub/EPdata/K562/ENCFF621AIY/pc1.compart",
                    ]
structureWeightList = [0.5,1.2,5,2,2
                       
tichr_obj_noquantile.weightStructure(structureTypeList,structureFileList,structureWeightList)
tichr_obj_noquantile.computeAllGene("hic",ifUseHiCRef=False)

predictdfname= "combine5"
tichr_obj_noquantile.RgxDf.to_csv("resultdf_Fulco_2025/"+predictdfname+"_Rgx_beforematch.tsv",
                                  index=None,sep="\t")
tichr_obj_noquantile.RgDF.to_csv("resultdf_Fulco_2025/"+predictdfname+"_Rg_beforematch.tsv",
                                  index=None,sep="\t")

```

Matching truelabel and plot

The golddf need to be the following style:

- site chr
- site start
- site end
- gene ID (goldcol)
- True or False to be a real site-to-gene links (truecol)


``` python
rgxdir="resultdf_Fulco_2025/"
golddf = "/home/wang/Tichr/datahub/EPinter/EP_CRISPRi_screens/hg38/Fulco_interactions_wScore.txt"
outdir = "K562fulco/"
tpmfile="/home/wang/Tichr/datahub/EPinter/K562_NT_H3K27ac+DNase/RNAseq/Matrix_edgeR/K562.genes.TPM.txt"


compareAdjust(rgxdir,golddf,outdir,tpmfile,tpmcol=[2,3],goldcol=4,title="K562 Fulco Dataset",
             if1stmatch=False,adjtype="sumrank")

```

<img src="_static/adjrgx/001.png" style="zoom:80%;" />

</details>







