# 5. Regulation-transcription concordance analysis

> While many factors act as general transcriptional regulators, they do not necessarily contribute to **treatment-induced** differential expression.

To evaluate this, we compared two types of correlations:  
1. **Corr(Rg, TPM)** — capturing the relationship between overall regulatory strength and basal gene expression.  
2. **Corr(ΔRg, logFC)** — measuring how regulatory *changes* correspond to transcriptional *changes* upon treatment.

A larger Corr(ΔRg, logFC) indicates a stronger treatment-specific regulatory contribution.

---

## Input requirements
The input consists of **paired files** (`Control` and `Treatment`) following a consistent filename pattern.  
Each **Control–Treatment pair** represents one experimental comparison.

### Required Columns 

Each RgDF file must contain the following columns:

| Column | Description |
|--------|-------------|
| **1** | Gene chromosome |
| **2** | Gene start position |
| **3** | Gene end position |
| **4** | Gene ID |
| **5** | Gene symbol |
| **6** | Gene strand |
| **7** | Gene fold change (log₂FC of treatment vs. control) |
| **8** | Gene FDR |
| **9** | Gene TPM (average of treatment and control) |
| **10** | Gene Rg score |

All columns should be **tab-separated**, and genomic coordinates must match the reference genome used during Rg computation.

For more information about how to generate this RgDF, please check [here](5.Negative.md#1-prepare-the-input-data)


### **Format Example (`*_RP_RgDf.tsv`):**

```
chr21	6111130	6123778	ENSG00000275993	ENSG00000275993	+	-1.73864248825759	7.784942363362921e-09	7.07768626695491	0.0
chr21	17513129	17593579	ENSG00000154639	CXADR	+	2.04164367091244	9.25684832742846e-09	5.14892581358436	4579.490782097491
```


---

## Prepare data

We provide a set of example data to help you get started quickly.

Specify the list of treatment–control pairs you want to compare:

```python
from tichr.treatment import *

datadir = "../Data/Treatment/"

pair_list = [
    ['siNIPBL_AFF4_rep0','Control_AFF4_rep0'],
    ['siNIPBL_H3K27ac_rep0','Control_H3K27ac_rep0'],
    ['siNIPBL_H3K27ac_rep2','Control_H3K27ac_rep2'],
    ['siNIPBL_H3K27me3_rep0','Control_H3K27me3_rep0'],
    ['siNIPBL_H3K27me3_rep2','Control_H3K27me3_rep2'],
    ['siNIPBL_H3K36me3_rep0','Control_H3K36me3_rep0'],
    ['siNIPBL_H3K36me3_rep2','Control_H3K36me3_rep2'],
    ['siNIPBL_H3K4me2_rep0','Control_H3K4me2_rep0'],
    ['siNIPBL_H3K4me3_rep0','Control_H3K4me3_rep0'],
    ['siNIPBL_Mau2_rep0','Control_Mau2_rep1'],
    ['siNIPBL_Pol2_rep0','Control_Pol2_rep0'],
    ['siNIPBL_Pol2Ser2_rep0','Control_Pol2Ser2_rep0'],
    ['siNIPBL_Rad21_rep0','Control_Rad21_rep1'],
    ['siNIPBL_Rad21_rep2','Control_Rad21_rep2'],
    ['JQ1plus_BRD4_rep0','JQ1minus_BRD4_rep0'],
]
```

Also, define a corresponding list of labels for plotting:

```python
factorlist = [
    "AFF4","H3K27ac","H3K27ac","H3K27me3","H3K27me3","H3K36me3","H3K36me3",
    "H3K4me2","H3K4me3","Mau2","Pol2","Pol2Ser2","Rad21","Rad21","BRD4"
]
```

---

## Run correlation analysis

Use the built-in `corrfunc` function to compute the correlation values for all defined expriment pairs:

```python
corMean, corTpm= corrfunc(datadir,pair_list, corrtype="pearson",suffix="_RgDf.tsv")
```

- **corMean**: correlation between mean Rg and mean TPM  
- **corTpm**: correlation between ΔRg and gene logFC  

You may also set `corrtype="spearman"` to compute Spearman correlation.



## Result Visualization

The function `plot_correlation_scatter` creates a scatter plot of the results:

```python
plot_correlation_scatter(corrrg, corrdelta, factorlist, title="RPE_siNIPBL")
```

<img src="_static/treatment/001.png" style="zoom:80%;" />

- **X-axis**: correlation between Rg and TPM  
- **Y-axis**: correlation between ΔRg and gene logFC  
- **Labels**: experimental marks (e.g., H3K27ac, H3K4me3, BRD4)  

**Upper part of the plot**: marks where ΔRg and ΔTPM are strongly positively correlated, and their changes (ΔRg) are consistent with gene expression changes (**treatment-associated regulation**).  

**Right part of the plot**: marks where Rg and TPM are strongly positively correlated (**general regulation**).  



## Regulatory concordance analysis

### Prepare Rg and RgX file of two conditions

``` python
from tichr.context import *

datadir="./Data/Negative"

rgCtrl_path = f"{datadir}/Ctrl_RgDf.tsv"
rgTreat_path = f"{datadir}/Treat_RgDf.tsv"
rgxCtrl_path = f"{datadir}/Ctrl_RgxDf.tsv"
rgxTreat_path = f"{datadir}/Treat_RgxDf.tsv"
```

The input file format is the same as [here](5.Negative.md#1-prepare-the-input-data)

### Combine treat and ctrl

``` python
rg_merged,rgx_merged = mergeDF(rgCtrl_path,rgTreat_path,rgxCtrl_path,rgxTreat_path,
                               minRgx=0.01,minRgxRatio=0.01,minRgxQuantile=0.01)
```

- `minRgx`: filter the site-to-gene links by `RgX value > min Rgx`

- `minRgxQuantile`: filter the site-to-gene links by `Quantile(RgX value) > minRgxQuantile`

- `minRgxRatio`: filter the site-to-gene links by `RgX Ratio > min RgxRatio`

The returned `rgx_merged` is a pandas DataFrame object containing the following columns:

- column12: RgX score for Ctrl
- column13: RgX ratio, averaged by Ctrl and Treat
- column14: RgX score for Treat

The returned `rg_merged` is a pandas DataFrame object containing the following columns:

- column10: Rg score for Ctrl
- column11: Rg score for Treat


### Compare the concordance

``` python
prepare_select_by_rank(rg_merged, basedon="rg",filetype="pandas",                        
                       title="MCF7 with estrogen",label="ER",outprefix="test")
```

- `basedon`: could be "rg" or "deltarg". For example, if rg is selected, the program will selected genes with highly related Rg-TPM, and examine if the changes of them， i.e., ΔRg-logFC(TPM) ， are also related.

- `filetype`: could be "pandas" or "file", for the format of input `rg_merged`. If file, a tab-separated without head is needed.


There are mainly three parts of output

1. Scatter plot. The scatter plot above illustrates the relationship between **Rg** and **TPM**, while the scatter plot below demonstrates the relationship between **ΔRg** and **logFC(TPM)**.

<img src="_static/negative/001.png" style="zoom:80%;" />

<img src="_static/negative/002.png" style="zoom:80%;" />

2. Select top correlated genes by Rg-TPM diffrank (left and middle), examine the diffrank of ΔRg-logFC(TPM)

<img src="_static/negative/003.png" style="zoom:80%;" />

3. Select top correlated genes by Rg-TPM sumrank (left and middle), examine the sumrank of ΔRg-logFC(TPM)

<img src="_static/negative/004.png" style="zoom:80%;" />


You can also select by ΔRg-logFC(TPM)

``` python
prepare_select_by_rank(rg_merged, basedon="deltarg",filetype="pandas",                        
                       title="MCF7 with estrogen",label="ER",outprefix="test")
``` 


<img src="_static/negative/005.png" style="zoom:80%;" />

<img src="_static/negative/006.png" style="zoom:80%;" />

### Full parameters

```
prepare_select_by_rank(rg_merged, basedon = "rg",,filetype="pandas",
                        negative_cutoff=0.8,positive_cutoff=0.8,
                        title="title",label="TF",outname=None,plotscatter=True)
```


- **`basedon`**: Method used for assessing regulatory contribution.  
  Options:  
  - `'rg'` — use absolute Rg values  
  - `'deltarg'` — use ΔRg (treatment–control changes)

- **`filetype`**: Format of the input data.  
  Options: `'pandas'` or `'file'`

- **`negative_cutoff`**:  
  Threshold for selecting **negatively correlated genes**, based on `diffrank`.

- **`positive_cutoff`**:  
  Threshold for selecting **positively correlated genes**, based on `sumrank`.

- **`outname`**:  
  If provided, a PDF file with the output plots will be generated using this prefix.

- **`plotscatter`**:  
  Boolean (`True`/`False`). Whether to generate scatter plots.

- **`title`**:  
  Title for the generated figure(s).

- **`label`**:  
  Names of transcriptional factors (TFs) used for annotation.