# 6. Identify context-specific negative regulation

**Tichr** can be used to **identify negative regulation for context-specific regulators**.  
We provide the `context` module to conveniently implement related analyses.




## 1. Prepare the input data

### Descriprion

> **The regulatory concordance analysis and identification of negative regulation in TICHR rely on specific treatments, as we assume that changes in regulatory effects relative to expression provide stronger causal evidence than static observations.**

> However, negative S2G links may be incompletely detected if the treatment only partially affects factor binding; therefore, treatments directly targeting the studied factors (e.g., E2 for ER binding) are recommended.

A custom algorithm was implemented to extract S2G links with **repressive regulatory functions**, which requires two inputs:

1. **A gene-level matrix** summarizing the Rg and expression values under both control and treatment conditions.
2. **An S2G-level matrix** containing RgX values for both conditions.

These two matrices are jointly used to identify regions whose regulatory influence decreases gene expression in a condition-specific manner.

### Prepare input

Therefore, you need to calculate the **RgX** and **Rg** values before this section.  
Here are several important points to pay attention to:

1. **Requirements for calculating regulatory changes between two conditions**
   - Epigenomic signals **before and after treatment** are required.
   - Differentially expressed genes (DEGs) **before and after treatment** are required.
   - Hi-C contacts for both conditions are optional.  
     You may either:
     - use the same Hi-C dataset for both conditions, or  
     - use distance-based weighting methods instead.

2. **Prepare columns 1–9 as the input file `candidateGeneFile` for `tichr`**  
   To provide differential gene expression information, the `candidateGeneFile` (see [here](1.Compute.md#1-compute-rg-and-rgx)) should follow this format:
   - **column1:** gene chr  
   - **column2:** gene start  
   - **column3:** gene end  
   - **column4:** gene ID  
   - **column5:** gene symbol  
   - **column6:** gene strand  
   - **column7:** gene fold change  
   - **column8:** gene FDR  
   - **column9:** gene TPM (average of treatment and control)

3. The resulting `RgDf.tsv` contains an additional column:
   - **column10:** gene Rg score

4. The Rg file **must NOT include S2G-level regulatory adjustment**.  
   See details in [Adjust for RgX and Rg](1.Compute.md#adjust-for-rgx-and-rg).

5. The Rgx file is the standard `RgxDf` generated by the `compute` module.  
   See [here](1.Compute.md).


``` python
from tichr.context import *

datadir="../Data/Negative"

rgCtrl_path = f"{datadir}/Ctrl_RgDf.tsv"
rgTreat_path = f"{datadir}/Treat_RgDf.tsv"
rgxCtrl_path = f"{datadir}/Ctrl_RgxDf.tsv"
rgxTreat_path = f"{datadir}/Treat_RgxDf.tsv"
```

### Preprocess

Combine treated and control group data:

``` python
rg_merged,rgx_merged = mergeDF(rgCtrl_path,rgTreat_path,rgxCtrl_path,rgxTreat_path,
                               minRgx=0.01,minRgxRatio=0.01,minRgxQuantile=0.01)

```

- `minRgx`: filter the site-to-gene links by `RgX value > min Rgx`

- `minRgxQuantile`: filter the site-to-gene links by `Quantile(RgX value) > minRgxQuantile`

- `minRgxRatio`: filter the site-to-gene links by `RgX Ratio > min RgxRatio`


The returned `rgx_merged` is a pandas DataFrame object containing the following columns:

- column12: RgX score for Ctrl
- column13: RgX ratio, averaged by Ctrl and Treat
- column14: RgX score for Treat

The returned `rg_merged` is a pandas DataFrame object containing the following columns:

- column10: Rg score for Ctrl
- column11: Rg score for Treat



### Identify negative regulation

Tichr can identify negatively functional site-to-gene pairs through the following steps:

**The following provide a brief introduction. For accurate description, please check our original paper.**

> 1. **Data Integration & Baseline:**  A baseline Pearson correlation between gene-level regulatory signal (Rg) and gene expression (TPM) is computed.
> 2. **Candidate Selection:** Site-to-gene links are selected based on two criteria:
>   - The regulatory site itself shows significant change (|logFC| > 0.5, FDR < 0.1).
>   - The linked gene is differentially expressed (|logFC| > 1, FDR < 0.05).
> 3. **Identification of Negative Candidates:** From the selected links, those where the direction of change in regulation (ΔRgX) is opposite to the direction of change in gene expression (gene logFC) are identified as candidate negative links (i.e., ΔRgX * gene logFC < 0).
> 4. **Signal Recalculation & Validation:** For these candidate negative links:
>   - A modified regulatory signal (**RgX**) is calculated by inverting ΔRgX and weighting it by its rank difference with the gene's logFC.
>   - A new gene-level regulatory score (**Rg**) is generated by aggregating the recalculated RgX' values.
> 5. **Iterative Improvement:**
>   - The correlation between the new score (**Rg**) and TPM is compared to the original baseline.
>   - If correlation improves, the consistency of the rank difference (diffrank) for each individual gene is checked.
>   - If all genes show improved rank coherence, the process stops.
>   - If not, only the genes with improved diffrank are kept, the set of negative links is refined based on this subset, and the process (steps 4-5) repeats.
> 6. **Termination:** The iterative loop continues until no further improvement in the correlation is observed.

The rg_merged and rgx_merged are used:

``` python
outdir="NegativeResults"
extractNeg(rg_merged, rgx_merged,filetype="pandas",
           corrtype="pearson",showInteration=False,
           outdir=outdir,outname="ERnegative",
           geneFC_cutoff=0.5,geneFDR_cutoff=0.05,
           minRgxRatio=0.01, rgxFC_cutoff=0.5)
```

- **filetype**:  
  Can be `"pandas"` or `"file"`, indicating the format of the input `rg_merged` and `rgx_merged`.  
  If set to `"file"`, the input must be a **tab-separated file without a header**.

- **showInteration**:  
  Whether to display the results of each iteration.  
  Default = `False`.

- **corrtype**:  
  Correlation method, either `"spearman"` or `"pearson"`.  
  Default = `"pearson"`.

- **outdir**:  
  Output directory.

- **outname**:  
  Output file prefix.

- **minRgxRatio**:  
  Minimum RgX ratio required for defining negative S2G links.

- **geneFC_cutoff**:  
  Cutoff for the **absolute log fold-change** (|logFC|) of gene expression.

- **geneFDR_cutoff**:  
  Cutoff for the **FDR** of gene expression.

- **rgxFC_cutoff**:  
  Cutoff for the **absolute log fold-change** (|logFC|) of **gene-level regulation**.



### Expected output

1. Text information

``` 
pearson correlation - Before: 0.1094, After: 0.3264, Difference: 0.2170
ERnegative
Number of genes with negative sites: 1108
Percentage of genes  with negative sites: 0.2875681287308591
Number of negative sites: 7012
Percentage of negative sites for degs: 0.22255371822134765
```

2. Tab-separated output files for negtive S2G links. Those files are subset of rgx_merged
   - ERnegative_negSite.tsv
   - ERnegative_negSite_filterRgxRatio.tsv

3. Visualizations

- Final selected site-to-gene links. In the plot, the s2g links highlighted in color are those identified as context-specific:

<img src="_static/negative/007.png" style="zoom:80%;" />

- Adjusted links and genes:

<img src="_static/negative/008.png" style="zoom:80%;" />

- The adjustment for negative links can significanly improve the relationships between Rg-TPM and ΔRg-logFC(TPM)

<img src="_static/negative/009.png" style="zoom:80%;" />

## Identify with Replicated Epigenome Data

If you have replicated epigenomic experimental data, you can use the following example workflow to perform the analysis in this section. 
- This approach allows you to integrate replicate-level epigenome signals, compute condition-specific regulation, and identify negative S2G links with higher robustness.
- The differential analysis of epigenome changes are performed with DeSeq2, thus you need to install it in your R environment.

``` python
from tichr.context import *

folder="./Data/Negative/resultdf_shYY1_hic/"
shCtrl_rep1_Rg = folder+"ATAC-seq_EPSC_shCtrl_None_rep1_RP_RgDf.tsv"
shCtrl_rep1_Rgx = folder+"ATAC-seq_EPSC_shCtrl_None_rep1_RP_RgxDf.tsv"
shCtrl_rep2_Rg = folder+"ATAC-seq_EPSC_shCtrl_None_rep2_RP_RgDf.tsv"
shCtrl_rep2_Rgx = folder+"ATAC-seq_EPSC_shCtrl_None_rep2_RP_RgxDf.tsv"

shYy1_rep1_Rg = folder+"ATAC-seq_EPSC_shYy1_None_rep1_RP_RgDf.tsv"
shYy1_rep1_Rgx= folder+"ATAC-seq_EPSC_shYy1_None_rep1_RP_RgxDf.tsv"
shYy1_rep2_Rg = folder+"ATAC-seq_EPSC_shYy1_None_rep2_RP_RgDf.tsv"
shYy1_rep2_Rgx = folder+"ATAC-seq_EPSC_shYy1_None_rep2_RP_RgxDf.tsv"

rgCtrl_file = [shCtrl_rep1_Rg,shCtrl_rep2_Rg,]
rgTreat_file = [shYy1_rep1_Rg,shYy1_rep2_Rg,]
rgxCtrl_file = [shCtrl_rep1_Rgx,shCtrl_rep2_Rgx]
rgxTreat_file = [shYy1_rep1_Rgx,shYy1_rep2_Rgx]

rg_merged,rgx_merged = mergeDFmany(rgCtrl_file,rgTreat_file,rgxCtrl_file,rgxTreat_file,
                                 minRgx=0.1,minRgxRatio=0.01,minRgxQuantile=0.01)


extractNeg(rg_merged, rgx_merged,filetype="pandas",
           corrtype="pearson",showInteration=False,
           outdir=outdir,outname="ERnegative",
           geneFC_cutoff=0.5,geneFDR_cutoff=0.05,
           rgxFC_cutoff=0.5,epifdr=True, rgxFDR_cutoff=0.05)

```