# 5. Identify context-specific (negative) regulation

## Prepare Rg and RgX file of two conditions

``` python
import sys
sys.path.append('/home/wang/github/Tichr-CLI/tichr')
from tichr import *
from context import *

rgCtrl_path = "negative/Ctrl_RgDf.tsv"
rgTreat_path = "negative/Treat_RgDf.tsv"
rgxCtrl_path = "negative/Ctrl_RgxDf.tsv"
rgxTreat_path = "negative/Treat_RgxDf.tsv"
```

File requirement for RgDf. You can prepare column1~9 as the input file of tichr `candidateGeneFile`.
- column1: gene chr
- column2: gene star
- column3: gene end
- column4: gene ID
- column5: gene symbol
- column6: gene strand
- column7: gene Fold change
- column8: gene FDR
- column9: gene TPM (average of treat and ctrl)
- column10: gene Rg score

File requirment for RgxDf
- column1: site chr
- column2: site start
- column3: site end
- column4: site epigenome signal
- column5: gene ID
- column6: gene chr
- column7: gene start
- column8: gene end
- column9: gene strand
- column10: gene symbol
- column11: site-to-gene weight
- column12: Rgx score
- column13: Rgx ratio

Combine treat and ctrl

``` python
rg_merged,rgx_merged = mergeDF(rgCtrl_path,rgTreat_path,rgxCtrl_path,rgxTreat_path,minRgx=0.1,minRgxRatio=0.01)
```
`minRgx`: filter the site-to-gene links by RgX value > minRgx
`minRgxRatio`: filter the site-to-gene links by RgX Ratio > minRgxRatio

the obtained rgx_merged is:
- column12: RgX score for Ctrl
- column13: RgX ratio, averaged by Ctrl and Treat
- column14: RgX score for Treat

the obtained rg_merged is:
- column10: Rg score for Ctrl
- column11: Rg score for Treat



## Identify negatively functional site-to-gene pairs

To identify potentially negative regulatory site-to-gene links, we integrated differential chromatin regulation data (TiChR results) and gene expression (DEG) analyses under two conditions. First, we computed the Pearson correlation between gene-level regulatory signal (Rg) and gene expression (TPM) as a baseline. We then selected site-to-gene links showing significant changes (|logFC| > 0.5, FDR < 0.1) and connected to differentially expressed genes (|logFC| > 1, FDR < 0.05). Candidate negative links were defined as those where the direction of regulation and gene expression change was opposite (i.e., the product of ΔRgX and gene logFC < 0). For these links, we recalculated a modified regulatory signal (RgX') by inverting ΔRgX and weighting it by the rank difference between ΔRgX and gene logFC. A new gene-level regulatory score (Rg') was obtained by aggregating RgX'. If the correlation between Rg' and TPM improved compared to the original Rg, we further evaluated whether the diffrank between Rg' and TPM increased for each gene. If all genes showed improved rank coherence, the iteration was terminated. Otherwise, only genes with improved diffrank were retained for the next round, and the negative link set was refined accordingly. The process iterated until no further improvement was observed.

The rg_merged and rgx_merged are used.

``` python
extractNeg(rg_merged, rgx_merged,showInteration=False,
           corrtype="pearson",filetype="pandas",
           outdir="negative",outname="ERnegative",minRgxRatio=0.01)
```
`filetype`:  could be "pandas" or "file", for the format of input `rg_merged` and `rgx_merged`. If file, a tab-separated without head is needed.
`showInteration`: if show the results of each Interation. Default=False
`corrtype`: could be "spearman" or "pearson". Default="pearson"
`outdir`: output directory 
`outname`: output file prefix
`minRgxRatio`: minimal Rgx ratio for negative s2g links.

There are serverl outout:

1. text information
``` 
pearson correlation - Before: 0.1094, After: 0.3264, Difference: 0.2170
ERnegative
Number of genes with negative sites: 1108
Percentage of genes  with negative sites: 0.2875681287308591
Number of negative sites: 7012
Percentage of negative sites for degs: 0.22255371822134765
```

2. tab-separated output files for negtive s2g links. Those files are subset of rgx_merged
    - ERnegative_negSite.tsv
    - ERnegative_negSite_filterRgxRatio.tsv

3. Visualizations

3.1 final selected site-to-gene links

<img src="_static/negative/007.png" style="zoom:80%;" />

3.2 adjusted links and genes

<img src="_static/negative/008.png" style="zoom:80%;" />

3.3 The adjustment for negative links can significanly improve the relationships between Rg-TPM and ΔRg-logFC(TPM)

<img src="_static/negative/009.png" style="zoom:80%;" />


## If you have replicates for epigenomes
``` python
folder="~/Tichr/2024Oct-summary/ContextSpecific/EPSC_shYY1_denovo/resultdf_shYY1_hic/"
shCtrl_rep1_Rg = folder+"ATAC-seq_EPSC_shCtrl_None_rep1_RP_RgDf.tsv"
shCtrl_rep1_Rgx = folder+"ATAC-seq_EPSC_shCtrl_None_rep1_RP_RgxDf.tsv"
shCtrl_rep2_Rg = folder+"ATAC-seq_EPSC_shCtrl_None_rep2_RP_RgDf.tsv"
shCtrl_rep2_Rgx = folder+"ATAC-seq_EPSC_shCtrl_None_rep2_RP_RgxDf.tsv"
shYy1_rep1_Rg = folder+"ATAC-seq_EPSC_shYy1_None_rep1_RP_RgDf.tsv"
shYy1_rep1_Rgx= folder+"ATAC-seq_EPSC_shYy1_None_rep1_RP_RgxDf.tsv"
shYy1_rep2_Rg = folder+"ATAC-seq_EPSC_shYy1_None_rep2_RP_RgDf.tsv"
shYy1_rep2_Rgx = folder+"ATAC-seq_EPSC_shYy1_None_rep2_RP_RgxDf.tsv"

rgCtrl_file = [shCtrl_rep1_Rg,shCtrl_rep2_Rg,]
rgTreat_file = [shYy1_rep1_Rg,shYy1_rep2_Rg,]
rgxCtrl_file = [shCtrl_rep1_Rgx,shCtrl_rep2_Rgx]
rgxTreat_file = [shYy1_rep1_Rgx,shYy1_rep2_Rgx]

rg_merged,rgx_merged = mergeDFmany(rgCtrl_file,rgTreat_file,rgxCtrl_file,rgxTreat_file,minRgx=0.1,minRgxRatio=0.01)

extractNeg(rg_merged, rgx_merged,showInteration=False,
           corrtype="pearson",filetype="pandas",
           outdir="negative",outname="ERnegative",minRgxRatio=0.01,
           epifdr=True)
           
```