# 7. Time-series multiomics analysis

Tichr can also be used to analyze time-series data. In this tutorial, I will demonstrate how to use Tichr's `timeSeries` module to perform conflict gene analysis, differentially expressed gene analysis, and regulatory sequence analysis for time-series epigenome and 3D genome data at single time points, dual time points, and multiple time points.

Below I will demonstrate how to use the Tichr API for time-series data analysis on an example dataset.

## Single Gene Analysis

```python
import sys
sys.path.append('/home/sunpx/my_project/TichrTest/tichr')
from tichr import *
from timeSeries import *

datadir="/home/sunpx/my_project/TichrTest/Data/TimeSeries"

GEM_RgxDfs = [
    f"{datadir}/GEM/GEM_Tichr_Rgx.tsv",
    f"{datadir}/GEM/GEN_Epi*MeanHiC_Rgx.tsv",
    f"{datadir}/GEM/GEN_HiC*MeanEpi_Rgx.tsv",
    f"{datadir}/GEM/GEM_Expon_Rgx.tsv"
]
GEM_Rgx_lables = [
    "GEM Rgx",
    "GEM Rgx(avgHiC)",
    "GEM Rgx(avgEpi)",
    "GEM Rgx(Expon)"
]
```

File requirment for Time-Series RgxDf:

- column1: site chr
- column2: site start
- column3: site end
- column4: site epigenome signal
- column5: gene ID
- column6: gene chr
- column7: gene start
- column8: gene end
- column9: gene strand
- column10: gene symbol
- column11~(10+t): Rgx value for each time points
- column11+t: Target mark(N/Target)

You can display the trend of all Rgx for the target gene over time by the following codes, the target SiteToGene Pair will be highlighted:

```python
timepoint_counts = 8
outdir = "TimeSeries_GEM"
import os
if not os.path.exists(outdir):
    os.makedirs(outdir)

for file_path, ylabel in zip(GEM_RgxDfs, GEM_Rgx_lables):
    title = "GEM " + ylabel
    plotOneGene(file_path, timepoint_counts, title, outdir=outdir)
```

<img src="_static/timeseries/001.png" style="zoom:80%;" />

<img src="_static/timeseries/002.png" style="zoom:80%;" />

<img src="_static/timeseries/003.png" style="zoom:80%;" />

<img src="_static/timeseries/004.png" style="zoom:80%;" />





## Two Timepoints Analysis

#### Prepare package and data

```python
import sys
sys.path.append('/home/sunpx/my_project/TichrTest/tichr')
from tichr import *
from timeSeries import *

datadir="/home/sunpx/my_project/TichrTest/Data/TimeSeries"
RgDfs = [
    f"{datadir}/TPH-1_H3K27ac_HiC/t0h_RP_RgDf.tsv",
    f"{datadir}/TPH-1_H3K27ac_HiC/t0.5h_RP_RgDf.tsv",
    f"{datadir}/TPH-1_H3K27ac_HiC/t1h_RP_RgDf.tsv",
    f"{datadir}/TPH-1_H3K27ac_HiC/t1.5h_RP_RgDf.tsv",
    f"{datadir}/TPH-1_H3K27ac_HiC/t2h_RP_RgDf.tsv",
    f"{datadir}/TPH-1_H3K27ac_HiC/t4h_RP_RgDf.tsv",
    f"{datadir}/TPH-1_H3K27ac_HiC/t6h_RP_RgDf.tsv",
    f"{datadir}/TPH-1_H3K27ac_HiC/t24h_RP_RgDf.tsv",
]

RgxDfs = [
    f"{datadir}/TPH-1_H3K27ac_HiC/t0h_RP_RgxDf.tsv",
    f"{datadir}/TPH-1_H3K27ac_HiC/t0.5h_RP_RgxDf.tsv",
    f"{datadir}/TPH-1_H3K27ac_HiC/t1h_RP_RgxDf.tsv",
    f"{datadir}/TPH-1_H3K27ac_HiC/t1.5h_RP_RgxDf.tsv",
    f"{datadir}/TPH-1_H3K27ac_HiC/t2h_RP_RgxDf.tsv",
    f"{datadir}/TPH-1_H3K27ac_HiC/t4h_RP_RgxDf.tsv",
    f"{datadir}/TPH-1_H3K27ac_HiC/t6h_RP_RgxDf.tsv",
    f"{datadir}/TPH-1_H3K27ac_HiC/t24h_RP_RgxDf.tsv",
]

#This is an optional feature to provide additional epigenetic data. 
#Here, we have incorporated the ​​epigenome signal of promoter regions​​.
promoter_twopoints_Df = f"{datadir}/TPH-1_H3K27ac_HiC/Promoter_EpigenomeSignal_TwoPoints.tsv"
promoter_eightpoints_Df = f"{datadir}/TPH-1_H3K27ac_HiC/Promoter_EpigenomeSignal_EightPoints.tsv"
```

The Rg and Rgx files are the `RgDf` and `RgxDf` obtained through Tichr `compute`.

The sample data for the subsequent **Multiple Timepoints Analysis** is the same as the data prepared here.

#### Compute LFC

Tichr provides an API to calculate the Log2 Fold Change (LFC) between the initial and final time points, enabling simultaneous computation of LFC for

1. Enhancer Epigenome Activity
2. Chromatin Contact Strength
3. Rgx Raw Value

Here is an example:

```python
resdf = lfc_two_timepoints(RgxDfs[0], RgxDfs[-1], extra_df=promoter_twopoints_Df, extra_name="Promoter")
print(resdf.head(3))
```

| geneChr | geneStart | geneEnd | geneStrand |              geneID | geneSymbol | EnhancerLFC | ContactLFC |    RgxLFC | PromoterLFC |
| ------: | --------: | ------: | ---------: | ------------------: | ---------: | ----------: | ---------: | --------: | ----------: |
|    chr1 |     69091 |   70008 |          + | ENSG00000186092.4_1 |      OR4F5 |    0.363534 |   0.742941 | -0.394892 |   -0.955160 |
|    chr1 |    367640 |  368634 |          + |   ENSG00000235249.1 |     OR4F29 |    0.282767 |        NaN |       NaN |   -0.295193 |
|    chr1 |    621096 |  622034 |          - | ENSG00000273547.1_2 |     OR4F16 |    0.240425 |   1.624994 |  1.985691 |   -1.201738 |

#### Change Or Not Genes Analysis

Choose an threshold, use `change_or_not` function to filter changed genes and no change genes.This can be easily achieved through the following codes:

```python
names = [ "Promoter", "Enhancer", "Contact"]
change_marks = [True, False, False]
thresholds = [0.15, 0.15, 0.15]
outdir="TimeSeries_TwoPoints"

promoter_change_df = change_or_not(resdf, names, change_marks, thresholds=thresholds, plot_it=True, ylim=(-2.5,2.5), outdir=outdir)
```

- `resdf`: pandas data frame,  has been obtained in previous steps.
- `names`: names for target features.
- `change_marks`: A list for change marks. True for changed，False for no-change.
- `thresholds`: threshold for determining change or not.
- `plot_it`: optional. plot a figure to show the results. (Default: False)
- `outdir`: optional. direction to save figure. (Default: None)
- `ylim`: optional. y axis limitation. (Default: [-2, 2])

<img src="_static/timeseries/005.png" style="zoom:80%;" />

#### Conflict Genes Analysis

You can filter conflict genes by  `two_conflict`  function. Here is an example:

```python
names = [ "Promoter", "Enhancer", "Contact"]
change_marks = ["Up", "Down", "NoChange"]
thresholds = [0.1, 0.1, 0.1]
outdir="TimeSeries_TwoPoints"

promoter_change_df = two_conflict(resdf, names, change_marks, thresholds=thresholds, plot_it=True, outdir=outdir)
```

- `resdf`: pandas data frame,  has been obtained in previous steps.
- `names`: names for target features.
- `change_marks`: A list for change marks. "Up"for up-regulated，'Down' for down-regulated, 'NoChange' for no-change.
- `thresholds`: threshold for determining regulated or not.
- `plot_it`: optional. plot a figure to show the results. (Default: False)
- `outdir`: optional. direction to save figure. (Default: None)
- `ylim`: optional. y axis limitation. (Default: [-2, 2])

<img src="_static/timeseries/006.png" style="zoom:80%;" />





## Multiple Timepoints Analysis

#### Prepare data

Data has been prepared in previous steps.

#### Compute LFC

Tichr provides an API to simply merge different time points RgxDf, compute Enhancer, Contact, Rg value, and compute LFC value as output.

You can achieve it by following codes:

```python
lfc_df, lfc_cols_list = lfc_multi_timepoints(RgxDfs, extra_file=promoter_eightpoints_Df, extra_name="Promoter")
print(lfc_df.columns.tolist())
```

This function will return a pandas `DataFrame` object containing LFC (Log Fold Change) values for each feature at each time point, along with a list of column names corresponding to the LFC values. 

These outputs will be used in subsequent steps. You can also directly use this `DataFrame` for personalized analysis.

```
['geneChr', 'geneStart', 'geneEnd', 'geneStrand', 'geneID', 'geneSymbol', 'Enhancer_LFC_T2', 'Enhancer_LFC_T3', 'Enhancer_LFC_T4', 'Enhancer_LFC_T5', 'Enhancer_LFC_T6', 'Enhancer_LFC_T7', 'Enhancer_LFC_T8', 'Contact_LFC_T2', 'Contact_LFC_T3', 'Contact_LFC_T4', 'Contact_LFC_T5', 'Contact_LFC_T6', 'Contact_LFC_T7', 'Contact_LFC_T8', 'Rgx_LFC_T2', 'Rgx_LFC_T3', 'Rgx_LFC_T4', 'Rgx_LFC_T5', 'Rgx_LFC_T6', 'Rgx_LFC_T7', 'Rgx_LFC_T8', 'Promoter_LFC_T2', 'Promoter_LFC_T3', 'Promoter_LFC_T4', 'Promoter_LFC_T5', 'Promoter_LFC_T6', 'Promoter_LFC_T7', 'Promoter_LFC_T8']
```

- `RgxDfs`: a list of RgxDf files
- `extra_file`: optional. Use when you have data from other sources. Contains time-series data records. (Default: None)
- `extra_name`: optional. Feature name for extra file. (Defalule: None)

#### Compute STS

Steady Total Strength(STS) is an indicator for assessing the overall Regulatory Potential (RP) strength of time-series data. Tichr provides an API `steady_total_strength` to calculate this metric. Here is an example:

```python
features = ["Enhancer", "Contact", "Rgx", "Promoter"]
sts_df = compute_sts(lfc_df, features, lfc_cols_list)
print(sts_df.columns.tolist())
```

This function returns a pandas DataFrame object containing STS values. It can be used as input for `multi_points_change` or for other customized analyses.

```
['geneChr', 'geneStart', 'geneEnd', 'geneStrand', 'geneID', 'geneSymbol', 'Enhancer_LFC_T2', 'Enhancer_LFC_T3', 'Enhancer_LFC_T4', 'Enhancer_LFC_T5', 'Enhancer_LFC_T6', 'Enhancer_LFC_T7', 'Enhancer_LFC_T8', 'Contact_LFC_T2', 'Contact_LFC_T3', 'Contact_LFC_T4', 'Contact_LFC_T5', 'Contact_LFC_T6', 'Contact_LFC_T7', 'Contact_LFC_T8', 'Rgx_LFC_T2', 'Rgx_LFC_T3', 'Rgx_LFC_T4', 'Rgx_LFC_T5', 'Rgx_LFC_T6', 'Rgx_LFC_T7', 'Rgx_LFC_T8', 'Promoter_LFC_T2', 'Promoter_LFC_T3', 'Promoter_LFC_T4', 'Promoter_LFC_T5', 'Promoter_LFC_T6', 'Promoter_LFC_T7', 'Promoter_LFC_T8', 'Enhancer_STS', 'Contact_STS', 'Rgx_STS', 'Promoter_STS']
```

- `lfc_df`: get from `lfc_multi_timepoints` function
- `features`: a list of feature names
- `lfc_cols_list`: get from `lfc_multi_timepoints` function

#### Change Or Not Genes Analysis

You can filter changed genes and no-change genes by setting a threshold for the STS value.Tichr provides an API `multi_points_change` to facilitate convenient implementation of this functionality.

Here is an example to show Promoter-Change; Enhancer-NoChange; Contact NoChange genes.

```
features = [ "Promoter", "Enhancer", "Contact"]
change_marks = [True, False, False]
thresholds = [0.1, 0.1, 0.1]
times = ["T2", "T3", "T4", "T5", "T6", "T7", "T8"]
multi_points_change(testdf, features, change_marks, thresholds, times, plotit=True)  
```

<img src="_static/timeseries/007.png" style="zoom:80%;" />

#### Early Or Late Regulated Genes Analysis

Tichr can also be used to distinguish between Early Regulated Genes (ERGs) and Late Regulated Genes (LRGs). TSR (Total Strength Ratio) is a metric designed to evaluate the chronological order of changes in gene Rg values. Tichr provides an API to compute and visualize TSR:

```python
resdf = compute_tsr(RgxDfs, extra_file=promoter_eightpoints_Df, extra_name="Promoter", plotit=True, times=["T1", "T2", "T3", "T4", "T5", "T6", "T7", "T8"])
```

This function will return a pandas `DataFrame` object containing TSR values and output some images to demonstrate the classification performance of TSR：

<img src="_static/timeseries/008.png" style="zoom:80%;" />

<img src="_static/timeseries/009.png" style="zoom:80%;" />